modules/basics/src/main/java/com/opengamma/basics/currency/FxMatrix.java

```diff
@@ -0,0 +1,557 @@
+     *     are other currencies present in the builder.
+     *   </li>
+     * </ul>
+     * An invocation of the method with {@code builder.addRate(GBP, USD, 1.6)}
+     * indicates that 1 pound sterling is worth 1.6 US dollars. It is
+     * equivalent to: {@code builder.addRate(USD, GBP, 1 / 1.6)} (1 US dollar
+     * is worth 0.625 pounds sterling) for all cases except where the rates
+     * are being updated.
+     *
+     * @param ccy1  the first currency of the pair
+     * @param ccy2  the second currency of the pair
+     * @param rate  the FX rate between the first currency and the second currency.
+     *   The rate indicates the value of one unit of the first currency in terms
+     *   of the second currency.
+     * @return the builder updated with the new rate
+     */
+    public Builder addRate(Currency ccy1, Currency ccy2, double rate) {
+
+      ArgChecker.notNull(ccy1, "ccy1");
+      ArgChecker.notNull(ccy2, "ccy2");
+
+      if (currencies.isEmpty()) {
+        addInitialCurrencyPair(ccy1, ccy2, rate);
+      } else {
+        addCurrencyPair(ccy1, ccy2, rate);
+      }
+      return this;
+    }
+
+    /**
+     * Adds a collection of new rates for currency pairs to the builder.
+     * Pairs that are already in the builder are treated as updates to the
+     * existing rates
+     *
+     * @param rates  the currency pairs and rates to be added
+     * @return the builder updated with the new rates
+     */
+    public Builder addRates(Map<CurrencyPair, Double> rates) {
+
+      ArgChecker.notNull(rates, "rates");
+
+      if (!rates.isEmpty()) {
+
+        ensureCapacity(
+            rates.keySet()
+                .stream()
+                .flatMap(cp ->
+                    Stream.<Currency>of(cp.getBase(), cp.getCounter())));
+
+        rates.entrySet()
+            .stream()
+            .forEach(e -> addRate(e.getKey(), e.getValue()));
+      }
+
+      return this;
+    }
+
+    private Builder() {
+      this.currencies = new LinkedHashMap<>();
+      this.rates = new double[MINIMAL_MATRIX_SIZE][MINIMAL_MATRIX_SIZE];
+    }
+
+    private Builder(ImmutableMap<Currency, Integer> currencies, double[][] rates) {
+      this.currencies = new LinkedHashMap<>(currencies);
+      // Ensure there is space to add at least one new currency
+      this.rates = copyArray(rates, size(currencies.size() + 1));
+    }
+
+    private Builder merge(Builder other) {
+
+      // Find the common currencies
+      Optional<Currency> common = currencies.keySet()
+          .stream()
+          .filter(other.currencies::containsKey)
+          .findFirst();
+
+      if (!common.isPresent()) {
+        throw new IllegalArgumentException("There are no currencies in common between " +
+            currencies.keySet() + " and " + other.currencies.keySet());
+      }
+
+      Currency commonCurrency = common.get();
+
+      // Add in all currencies that we don't already have
+      other.currencies.entrySet()
```

Comment:
This might be neater if you used ``Sets.difference(other.currencies.entrySet(), currencies.entrySet())`` instead of doing the filter.